File: Tests/Tests/STPSourceReceiverTest.m
//
//  STPSourceReceiverTest.m
//  Stripe
//
//  Created by Joey Dong on 6/26/17.
//  Copyright © 2017 Stripe, Inc. All rights reserved.
//

@import XCTest;


#import "STPFixtures.h"
#import "STPTestUtils.h"

@interface STPSourceReceiverTest : XCTestCase

@end

@implementation STPSourceReceiverTest

#pragma mark - Description Tests

- (void)testDescription {
    STPSourceReceiver *receiver = [STPSourceReceiver decodedObjectFromAPIResponse:[STPTestUtils jsonNamed:STPTestJSONSource3DS][@"receiver"]];
    XCTAssert(receiver.description);
}

#pragma mark - STPAPIResponseDecodable Tests

- (void)testDecodedObjectFromAPIResponseRequiredFields {
    NSArray<NSString *> *requiredFields = @[
                                            @"address",
                                            ];

    for (NSString *field in requiredFields) {
        NSMutableDictionary *response = [[STPTestUtils jsonNamed:STPTestJSONSource3DS][@"receiver"] mutableCopy];
        [response removeObjectForKey:field];

        XCTAssertNil([STPSourceReceiver decodedObjectFromAPIResponse:response]);
    }

    XCTAssert([STPSourceReceiver decodedObjectFromAPIResponse:[STPTestUtils jsonNamed:STPTestJSONSource3DS][@"receiver"]]);
}

- (void)testDecodedObjectFromAPIResponseMapping {
    NSDictionary *response = [STPTestUtils jsonNamed:STPTestJSONSource3DS][@"receiver"];
    STPSourceReceiver *receiver = [STPSourceReceiver decodedObjectFromAPIResponse:response];

    XCTAssertEqualObjects(receiver.address, @"test_1MBhWS3uv4ynCfQXF3xQjJkzFPukr4K56N");
    XCTAssertEqualObjects(receiver.amountCharged, @(300));
    XCTAssertEqualObjects(receiver.amountReceived, @(200));
    XCTAssertEqualObjects(receiver.amountReturned, @(100));

    XCTAssertNotEqual(receiver.allResponseFields, response);
    XCTAssertEqualObjects(receiver.allResponseFields, response);
}

@end

File: Stripe/StripeiOSTests/STPSourceReceiverTest.swift
//  Converted to Swift 5.8.1 by Swiftify v5.8.28463 - https://swiftify.com/
//
//  STPSourceReceiverTest.m
//  Stripe
//
//  Created by Joey Dong on 6/26/17.
//  Copyright © 2017 Stripe, Inc. All rights reserved.
//

import XCTest

class STPSourceReceiverTest: XCTestCase {
    // MARK: - Description Tests

    func testDescription() {
        let receiver = STPSourceReceiver.decodedObject(fromAPIResponse: STPTestUtils.jsonNamed(STPTestJSONSource3DS)["receiver"] as? [AnyHashable: Any])
        XCTAssertNotNil(receiver?.description)
    }

    // MARK: - STPAPIResponseDecodable Tests

    func testDecodedObjectFromAPIResponseRequiredFields() {
        let requiredFields = [
            "address",
        ]

        for field in requiredFields {
            var response = STPTestUtils.jsonNamed(STPTestJSONSource3DS)["receiver"] as? [AnyHashable: Any]
            response?.removeValue(forKey: field)

            XCTAssertNil(STPSourceReceiver.decodedObject(fromAPIResponse: response))
        }

        XCTAssertNotNil(STPSourceReceiver.decodedObject(fromAPIResponse: STPTestUtils.jsonNamed(STPTestJSONSource3DS)["receiver"] as? [AnyHashable: Any]))
    }

    func testDecodedObjectFromAPIResponseMapping() {
        let response = STPTestUtils.jsonNamed(STPTestJSONSource3DS)["receiver"] as? [AnyHashable: Any]
        let receiver = STPSourceReceiver.decodedObject(fromAPIResponse: response)

        XCTAssertEqual(receiver?.address, "test_1MBhWS3uv4ynCfQXF3xQjJkzFPukr4K56N")
        XCTAssertEqual(receiver?.amountCharged, NSNumber(value: 300))
        XCTAssertEqual(receiver?.amountReceived, NSNumber(value: 200))
        XCTAssertEqual(receiver?.amountReturned, NSNumber(value: 100))

        XCTAssertEqual(receiver!.allResponseFields as NSDictionary, response! as NSDictionary)
    }
}
